test/zipkin/util.go

```diff
@@ -48,54 +47,74 @@ const (
 	ZipkinNamespace = "istio-system"
 )
 
-var zipkinPortForwardPID int
+var (
+	zipkinPortForwardPID int
+
+	// ZipkinTracingEnabled variable indicating if zipkin tracing is enabled.
+	ZipkinTracingEnabled = false
+
+	// sync.Once variable to ensure we execute zipkin setup only once.
+	setupOnce sync.Once
+
+	// sync.Once variable to ensure we execute zipkin cleanup only if zipkin is setup and it is executed only once.
+	teardownOnce sync.Once
+)
 
 // SetupZipkinTracing sets up zipkin tracing which involves:
 // 1. Setting up port-forwarding from localhost to zipkin pod on the cluster
 //    (pid of the process doing Port-Forward is stored in a global variable).
 // 2. Enable AlwaysSample config for tracing.
-func SetupZipkinTracing(kubeClientset *kubernetes.Clientset) error {
-	logger := logging.GetContextLogger("SpoofUtil")
-
-	if err := CheckZipkinPortAvailability(); err != nil {
-		return fmt.Errorf("Zipkin port not available on the machine: %v", err)
-	}
-
-	zipkinPods, err := kubeClientset.CoreV1().Pods(ZipkinNamespace).List(metav1.ListOptions{LabelSelector: "app=zipkin"})
-	if err != nil {
-		return fmt.Errorf("Error retrieving Zipkin pod details: %v", err)
-	}
-
-	if len(zipkinPods.Items) == 0 {
-		return errors.New("No Zipkin Pod found on the cluster. Ensure monitoring is switched on for your Knative Setup")
-	}
-
-	portForwardCmd := exec.Command("kubectl", "port-forward", "--namespace="+ZipkinNamespace, zipkinPods.Items[0].Name, fmt.Sprintf("%d:%d", ZipkinPort, ZipkinPort))
-	if err = portForwardCmd.Start(); err != nil {
-		return fmt.Errorf("Error starting kubectl port-forward command: %v", err)
-	}
-	zipkinPortForwardPID = portForwardCmd.Process.Pid
-	logger.Infof("Zipkin port-forward process started with PID: %d", zipkinPortForwardPID)
-
-	// Applying AlwaysSample config to ensure we propagate zipkin header for every request made by this client.
-	trace.ApplyConfig(trace.Config{DefaultSampler: trace.AlwaysSample()})
-
-	logger.Info("Successfully setup SpoofingClient for Zipkin Tracing")
-
-	return nil
+func SetupZipkinTracing(kubeClientset *kubernetes.Clientset) {
+	setupOnce.Do(func() {
+		logger := logging.GetContextLogger("SpoofUtil")
```

Comment:
we are using FormatLogger everywhere now instead of context logger - https://github.com/knative/pkg/blob/master/test/spoof/spoof.go#L79

We should use that here as well.

Reply:
How  do you instantiate a FormatLogger, all places I see in the code base its through Test. Can you please point me to the documentation on how to use it, also it should not be Spoof Client? it should be in the logging package with proper documentation.

Reply:
Maybe we can move it to logging. I am not sure if there is much documentation on it. You can check on existing tests on how to use it.